File: swift-bridge/Sources/SceneKitBridge/Physics.swift
import SceneKit

@_cdecl("scn_physics_body_static")
public func scn_physics_body_static() -> UnsafeMutableRawPointer? {
    scnRetain(SCNPhysicsBody.static())
}

@_cdecl("scn_physics_body_dynamic")
public func scn_physics_body_dynamic() -> UnsafeMutableRawPointer? {
    scnRetain(SCNPhysicsBody.dynamic())
}

@_cdecl("scn_physics_body_kinematic")
public func scn_physics_body_kinematic() -> UnsafeMutableRawPointer? {
    scnRetain(SCNPhysicsBody.kinematic())
}

@_cdecl("scn_physics_body_get_type")
public func scn_physics_body_get_type(_ bodyHandle: UnsafeMutableRawPointer?) -> Int32 {
    guard let body: SCNPhysicsBody = scnBorrow(bodyHandle) else { return -1 }
    return Int32(clamping: body.type.rawValue)
}

@_cdecl("scn_physics_body_set_type")
public func scn_physics_body_set_type(_ bodyHandle: UnsafeMutableRawPointer?, _ bodyType: Int32) {
    guard let body: SCNPhysicsBody = scnBorrow(bodyHandle) else { return }
    body.type = SCNPhysicsBodyType(rawValue: Int(bodyType)) ?? .dynamic
}

@_cdecl("scn_physics_body_get_mass")
public func scn_physics_body_get_mass(_ bodyHandle: UnsafeMutableRawPointer?) -> Double {
    guard let body: SCNPhysicsBody = scnBorrow(bodyHandle) else { return 0 }
    return body.mass
}

@_cdecl("scn_physics_body_set_mass")
public func scn_physics_body_set_mass(_ bodyHandle: UnsafeMutableRawPointer?, _ mass: Double) {
    guard let body: SCNPhysicsBody = scnBorrow(bodyHandle) else { return }
    body.mass = mass
}

@_cdecl("scn_physics_body_get_restitution")
public func scn_physics_body_get_restitution(_ bodyHandle: UnsafeMutableRawPointer?) -> Double {
    guard let body: SCNPhysicsBody = scnBorrow(bodyHandle) else { return 0 }
    return body.restitution
}

@_cdecl("scn_physics_body_set_restitution")
public func scn_physics_body_set_restitution(_ bodyHandle: UnsafeMutableRawPointer?, _ restitution: Double) {
    guard let body: SCNPhysicsBody = scnBorrow(bodyHandle) else { return }
    body.restitution = restitution
}

@_cdecl("scn_physics_body_get_friction")
public func scn_physics_body_get_friction(_ bodyHandle: UnsafeMutableRawPointer?) -> Double {
    guard let body: SCNPhysicsBody = scnBorrow(bodyHandle) else { return 0 }
    return body.friction
}

@_cdecl("scn_physics_body_set_friction")
public func scn_physics_body_set_friction(_ bodyHandle: UnsafeMutableRawPointer?, _ friction: Double) {
    guard let body: SCNPhysicsBody = scnBorrow(bodyHandle) else { return }
    body.friction = friction
}

@_cdecl("scn_physics_body_get_category_bit_mask")
public func scn_physics_body_get_category_bit_mask(_ bodyHandle: UnsafeMutableRawPointer?) -> UInt {
    guard let body: SCNPhysicsBody = scnBorrow(bodyHandle) else { return 0 }
    return UInt(bitPattern: body.categoryBitMask)
}

@_cdecl("scn_physics_body_set_category_bit_mask")
public func scn_physics_body_set_category_bit_mask(_ bodyHandle: UnsafeMutableRawPointer?, _ mask: UInt) {
    guard let body: SCNPhysicsBody = scnBorrow(bodyHandle) else { return }
    body.categoryBitMask = Int(bitPattern: mask)
}

@_cdecl("scn_physics_body_get_collision_bit_mask")
public func scn_physics_body_get_collision_bit_mask(_ bodyHandle: UnsafeMutableRawPointer?) -> UInt {
    guard let body: SCNPhysicsBody = scnBorrow(bodyHandle) else { return 0 }
    return UInt(bitPattern: body.collisionBitMask)
}

@_cdecl("scn_physics_body_set_collision_bit_mask")
public func scn_physics_body_set_collision_bit_mask(_ bodyHandle: UnsafeMutableRawPointer?, _ mask: UInt) {
    guard let body: SCNPhysicsBody = scnBorrow(bodyHandle) else { return }
    body.collisionBitMask = Int(bitPattern: mask)
}

@_cdecl("scn_physics_body_get_contact_test_bit_mask")
public func scn_physics_body_get_contact_test_bit_mask(_ bodyHandle: UnsafeMutableRawPointer?) -> UInt {
    guard let body: SCNPhysicsBody = scnBorrow(bodyHandle) else { return 0 }
    return UInt(bitPattern: body.contactTestBitMask)
}

@_cdecl("scn_physics_body_set_contact_test_bit_mask")
public func scn_physics_body_set_contact_test_bit_mask(_ bodyHandle: UnsafeMutableRawPointer?, _ mask: UInt) {
    guard let body: SCNPhysicsBody = scnBorrow(bodyHandle) else { return }
    body.contactTestBitMask = Int(bitPattern: mask)
}

@_cdecl("scn_physics_body_apply_force")
public func scn_physics_body_apply_force(
    _ bodyHandle: UnsafeMutableRawPointer?,
    _ x: Float,
    _ y: Float,
    _ z: Float,
    _ impulse: Bool
) {
    guard let body: SCNPhysicsBody = scnBorrow(bodyHandle) else { return }
    body.applyForce(
        SCNVector3(x: CGFloat(x), y: CGFloat(y), z: CGFloat(z)),
        asImpulse: impulse
    )
}

File: src/physics.rs
use crate::ffi;
use crate::math::Vector3;
use crate::private::handle_type;

handle_type!(PhysicsBody);

/// Mirrors `SCNPhysicsBodyType`.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[repr(i32)]
pub enum PhysicsBodyType {
    /// Corresponds to the `SCNPhysicsBodyType::Static` case.
    Static = 0,
    /// Corresponds to the `SCNPhysicsBodyType::Dynamic` case.
    Dynamic = 1,
    /// Corresponds to the `SCNPhysicsBodyType::Kinematic` case.
    Kinematic = 2,
}

impl PhysicsBodyType {
    const fn from_raw(value: i32) -> Option<Self> {
        match value {
            0 => Some(Self::Static),
            1 => Some(Self::Dynamic),
            2 => Some(Self::Kinematic),
            _ => None,
        }
    }
}

impl PhysicsBody {
    /// Mirrors `SCNPhysicsBody.staticBody`.
    #[must_use]
    pub fn static_body() -> Option<Self> {
        unsafe { Self::from_raw(ffi::scn_physics_body_static()) }
    }

    /// Mirrors `SCNPhysicsBody.dynamicBody`.
    #[must_use]
    pub fn dynamic_body() -> Option<Self> {
        unsafe { Self::from_raw(ffi::scn_physics_body_dynamic()) }
    }

    /// Mirrors `SCNPhysicsBody.kinematicBody`.
    #[must_use]
    pub fn kinematic_body() -> Option<Self> {
        unsafe { Self::from_raw(ffi::scn_physics_body_kinematic()) }
    }

    /// Mirrors `SCNPhysicsBody.bodyType`.
    #[must_use]
    pub fn body_type(&self) -> Option<PhysicsBodyType> {
        PhysicsBodyType::from_raw(unsafe { ffi::scn_physics_body_get_type(self.ptr) })
    }

    /// Sets the `SCNPhysicsBody.bodyType` member.
    pub fn set_body_type(&self, body_type: PhysicsBodyType) {
        unsafe { ffi::scn_physics_body_set_type(self.ptr, body_type as i32) };
    }

    /// Mirrors `SCNPhysicsBody.mass`.
    #[must_use]
    pub fn mass(&self) -> f64 {
        unsafe { ffi::scn_physics_body_get_mass(self.ptr) }
    }

    /// Sets the `SCNPhysicsBody.mass` member.
    pub fn set_mass(&self, mass: f64) {
        unsafe { ffi::scn_physics_body_set_mass(self.ptr, mass) };
    }

    /// Mirrors `SCNPhysicsBody.restitution`.
    #[must_use]
    pub fn restitution(&self) -> f64 {
        unsafe { ffi::scn_physics_body_get_restitution(self.ptr) }
    }

    /// Sets the `SCNPhysicsBody.restitution` member.
    pub fn set_restitution(&self, restitution: f64) {
        unsafe { ffi::scn_physics_body_set_restitution(self.ptr, restitution) };
    }

    /// Mirrors `SCNPhysicsBody.friction`.
    #[must_use]
    pub fn friction(&self) -> f64 {
        unsafe { ffi::scn_physics_body_get_friction(self.ptr) }
    }

    /// Sets the `SCNPhysicsBody.friction` member.
    pub fn set_friction(&self, friction: f64) {
        unsafe { ffi::scn_physics_body_set_friction(self.ptr, friction) };
    }

    /// Mirrors `SCNPhysicsBody.applyForce`.
    pub fn apply_force(&self, force: Vector3, impulse: bool) {
        unsafe { ffi::scn_physics_body_apply_force(self.ptr, force.x, force.y, force.z, impulse) };
    }
}
